index.js

```diff
@@ -70,54 +73,76 @@ module.exports = function (options) {
       origin = options.origin || requestOrigin;
     }
 
+    var headersSet = {};
+
+    function set(self, key, value) {
+      self.set(key, value);
+      headersSet[key] = value;
+    }
+
     if (this.method !== 'OPTIONS') {
       // Simple Cross-Origin Request, Actual Request, and Redirects
 
-      this.set('Access-Control-Allow-Origin', origin);
+      set(this, 'Access-Control-Allow-Origin', origin);
 
       if (options.credentials === true) {
-        this.set('Access-Control-Allow-Credentials', 'true');
+        set(this, 'Access-Control-Allow-Credentials', 'true');
       }
 
       if (options.exposeHeaders) {
-        this.set('Access-Control-Expose-Headers', options.exposeHeaders);
+        set(this, 'Access-Control-Expose-Headers', options.exposeHeaders);
       }
-
-      yield next;
     } else {
       // Preflight Request
 
       // If there is no Access-Control-Request-Method header or if parsing failed,
       // do not set any additional headers and terminate this set of steps.
       // The request is outside the scope of this specification.
-      if (!this.get('Access-Control-Request-Method')) {
```

Comment:
on `Preflight Request`, there is won't throw any error on old code logic https://github.com/koajs/cors/blob/master/index.js#L88.